dvc/render/utils.py

```diff
@@ -67,3 +100,94 @@ def render(
         template_path=html_template_path,
         refresh_seconds=refresh_seconds,
     )
+
+
+def _filter_fields(
+    datapoints: List[Dict], filename, revision, fields=None
+) -> List[Dict]:
+    if not fields:
+        return datapoints
+    assert isinstance(fields, set)
+
+    new_data = []
+    for data_point in datapoints:
+        new_dp = copy(data_point)
+
+        keys = set(data_point.keys())
+        if keys & fields != fields:
+            raise DvcException(
+                "Could not find fields: '{}' for '{}' at '{}'.".format(
+                    ", ".join(fields), filename, revision
+                )
+            )
+
+        to_del = keys - fields
+        for key in to_del:
+            del new_dp[key]
+        new_data.append(new_dp)
+    return new_data
+
+
+def _lists(dictionary: Dict):
+    for _, value in dictionary.items():
+        if isinstance(value, dict):
+            yield from _lists(value)
+        elif isinstance(value, list):
+            yield value
+
+
+def _find_data(data: Union[Dict, List], fields=None) -> List[Dict]:
+    if not isinstance(data, dict):
+        return data
+
+    if not fields:
+        # just look for first list of dicts
+        fields = set()
+
+    for lst in _lists(data):
+        if (
+            all(isinstance(dp, dict) for dp in lst)
+            and set(first(lst).keys()) & fields == fields
+        ):
+            return lst
+    raise PlotDataStructureError()
+
+
+def _append_index(datapoints: List[Dict], append_index=False) -> List[Dict]:
+    if not append_index or INDEX_FIELD in first(datapoints).keys():
+        return datapoints
+
+    for index, data_point in enumerate(datapoints):
+        data_point[INDEX_FIELD] = index
+    return datapoints
+
+
+def _append_revision(datapoints: List[Dict], revision) -> List[Dict]:
+    for data_point in datapoints:
+        data_point[REVISION_FIELD] = revision
+    return datapoints
+
+
+def to_datapoints(data: Dict, props: Dict):
```

Comment:
Is this the blessed preprocessor util, which one should import and use? It can't be used by Studio at least in this form:
- we need to skip `_append_revision()`
- need to fill in props `x` and `y` based on data

Reply:
Everything is so low-level and not really defined that I doubt it could be really reused.

Reply:
Ok, Ill modify it, how should

> need to fill in props x and y based on data

work?

Reply:
@Suor Can you clarify what you DO want this preprocessing util to do? I might be missing the context from previous conversations, but it's hard for me to tell from looking through the GitHub issues.

@pared Can we think about whether this would be useful when we start developing a public api to access plots data? For example, let's say I want to get all the data from one or multiple experiments to do my own analysis/visualization.

Reply:
@dberenbaum this is already in place: current `Repo.plots.show` will return mapping containing chosen targets from particular revisions. The change that we are doing here is supposed to convert this data returned from `Repo.plots.show` to datapoints which are a structure required by vega plots templates.

Reply:
@Suor sorry, should have checked `parse_data` before asking for review, I will look into moving the `x` and `y` resolution too. I will also think about how to alternate the structure so that those methods are not some randoms imported from `utils`.